src/poly1305.rs

```diff
@@ -16,31 +16,115 @@
 // TODO: enforce maximum input length.
 
 use {c, chacha, constant_time, error};
+use core;
+use polyfill::slice::u32_from_le_u8;
 
-pub struct SigningContext {
-    state: State,
+struct AlignedBlockState<'a> {
+    buf: [u8; BLOCK_STATE_SIZE + 7],
+    inner: &'a mut [u8; BLOCK_STATE_SIZE],
+}
+impl<'a> AlignedBlockState<'a> {
+    pub fn new(inner: &mut [u8; BLOCK_STATE_SIZE]) -> AlignedBlockState {
+        let mut st = AlignedBlockState {
+            buf: [0u8; BLOCK_STATE_SIZE + 7],
+            inner: inner,
+        };
+        AlignedBlockState::align(&mut st.buf).copy_from_slice(&st.inner[..]);
+        st
```

Comment:
This will construct an `AlignedBlockState` named `st`, calculate the offset, and then return a copy of it. But, the returned copy might not have the same alignment as `st`, so the offset might be wrong, right?